### ISSUE TITLE
return type inference

### ISSUE THREAD

**Issue by borkdude** [borkdude/clj-kondo#728]
- [x] warning:
``` clojure
(defn foo []
  :foo)

(inc (foo))
```
should give a warning.

- [x] fix false positive regression + write test:

```
(keyword (when-not false "foo") "bar")
```
=>
```
<stdin>:1:10: warning: Expected: string or nil, received: list.
```

- [x] function indirection:

```
(defn foo1 []
  :bar)

(defn foo2 []
  'baz)

(defn foo [x]
  (if (< x 5) (foo1) (foo2)))

(inc (foo 1))
```

- [x] don't store `{:type :call}` tags in the cache, but resolve them before writing! this requires some work in probably `ret-tag-from-call`

e.g.:

```
cat resources/clj_kondo/impl/cache/built_in/cljs/clojure.zip.transit.json | jet --from transit --pretty
```

- [x] recursive

```
(defn foo [] (foo)) (inc (foo))
```

- [x] stackoverflow

```
 clj -A:clj-kondo --lint $(cd /tmp/metabase; lein classpath) | grep Expected
```

```
insert-left {:row 165,
              :col 1,
              :fixed-arities #{2},
              :name insert-left,
              :ns clojure.zip,
              :top-ns clojure.zip,
              :arities {2 {:ret #{:vector
                                  {:call {:type :call,
                                          :lang :cljs,
                                          :base-lang :cljs,
                                          :resolved-ns cljs.core,
                                          :ns clojure.zip,
                                          :name throw,
                                          :arity 1},
                                   :row 171,
                                   :col 9,
                                   :end-row 171,
                                   :end-col 32}}}}},
```
